Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIDeclarativeNetRequestCocoa.mm

```diff
@@ -384,20 +390,26 @@
 {
     auto *ruleIDsToDelete = @[ ];
     if (ruleIDsToDeleteVector) {
-        ruleIDsToDelete = createNSArray(ruleIDsToDeleteVector.value(), [](double ruleID) {
+        ruleIDsToDelete = createNSArray(ruleIDsToDeleteVector.value(), [this, protectedThis = Ref { *this }](double ruleID) {
+            if (!m_sessionRulesIDs.contains(ruleID))
+                return (NSNumber *)nil;
             return @(ruleID);
         }).get();
     }
 
     NSArray *rulesToAdd = rulesToAddJSON ? parseJSON(rulesToAddJSON.value(), JSONOptions::FragmentsAllowed) : @[ ];
 
-    // FIXME: Make sure that adding these rules won't get us over the maximum number of dynamic + session rules.
-
     if (!ruleIDsToDelete.count && !rulesToAdd.count) {
         completionHandler(std::nullopt);
         return;
     }
 
+    double updatedSessionRulesCount = m_sessionRulesIDs.size() + rulesToAdd.count - ruleIDsToDelete.count;
+    if (updatedSessionRulesCount + m_dynamicRulesIDs.size() > webExtensionDeclarativeNetRequestMaximumNumberOfDynamicAndSessionRules) {
+        completionHandler(toErrorString(@"declarativeNetRequest.updateSessionRules()", nil, @"Failed to add session rules. Maximum number of dynamic and session rules exceeded."));
+        return;
+    }
+
     updateDeclarativeNetRequestRulesInStorage(declarativeNetRequestSessionRulesStore(), @"session", rulesToAdd, ruleIDsToDelete, WTFMove(completionHandler));
```

Comment:
Ditto on code that removes the deleted IDs from `m_sessionRulesIDs`.

Reply:
Ditto.